### PR TITLE
Sub labels hints for yolov8 pose

### PR DESCRIPTION
**Pull request by Eldies**
### Summary

Adding an option to YOLOv8 Pose extractor which allows to pass labels of points

### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/cvat-ai/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](https://github.com/cvat-ai/datumaro/tree/develop/docs) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 CVAT.ai Corporation
#
# SPDX-License-Identifier: MIT
```

## Summary by CodeRabbit

- **New Features**
  - Enhanced rectangle validation logic for improved accuracy in annotations.
  - Added support for skeleton sub-labels in the YOLOv8 pose extractor, improving flexibility in dataset configurations.

- **Bug Fixes**
  - Streamlined error messages for clarity when invalid annotations occur.
  - Added new tests to ensure proper handling of invalid shapes and sub-label configurations, enhancing error reporting.

- **Tests**
  - Introduced new test cases to validate the integrity of annotation imports and improve overall testing coverage.

### CODE REVIEW

File: tests/unit/data_formats/test_yolo_format.py [hunk 1]
```diff
@@ -1221,3 +1230,73 @@ def _make_some_annotation_values():
...
+                        "custom_name",
+                        "another_custom_name",
+                        "test_name",
+                        "42",
+                        "extra_sub_label",
+                    ],
+                },
+            )
+
+    def test_fails_on_lack_of_skeleton_name_in_hint(self, test_dir):
```
Comment (zhiltsov-max) [inline tests/unit/data_formats/test_yolo_format.py]: Please rename tests to use naming convention: `test_can_...`

Comment (Eldies) [reply tests/unit/data_formats/test_yolo_format.py]: fixed

File: datumaro/plugins/yolo_format/extractor.py [hunk 2]
```diff
@@ -334,25 +334,28 @@ def _config(self) -> Dict[str, Any]:
...
                 raise InvalidAnnotationError("Failed to parse config file")
 
-    def _load_categories(self) -> Dict[AnnotationType, LabelCategories]:
+    def _load_names_from_config_file(self):
+        names = self._config["names"]
+        if isinstance(names, dict):
+            if set(names.keys()) != set(range(len(names))):
+                raise InvalidAnnotationError(
+                    "Failed to parse names from config: non-sequential label ids"
+                )
```
Comment (zhiltsov-max) [inline datumaro/plugins/yolo_format/extractor.py]: What is the point of this check? If it's a dict, I think it should support arbitrary ordering.

Comment (Eldies) [reply datumaro/plugins/yolo_format/extractor.py]: YOLOv8 framework requires `n` class dataset to have class indices `0` - `n - 1`. 
https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/autobackend.py#L36

Do you think that datumaro should be able to import a dataset even if it does not meet the YOLOv8 requirements?

Comment (zhiltsov-max) [reply datumaro/plugins/yolo_format/extractor.py]: Yes, I think so. It looks like an unnecessary limitation for Datumaro.

Comment (Eldies) [reply datumaro/plugins/yolo_format/extractor.py]: Done

Comment (Eldies) [reply datumaro/plugins/yolo_format/extractor.py]: btw, may be we should also allow for string keys?

Comment (zhiltsov-max) [reply datumaro/plugins/yolo_format/extractor.py]: Do you mean arbitrary strings or strings with numbers? The latter one seems to be supported in the code by the link above, so it makes sense to support it in Datumaro as well.

Comment (Eldies) [reply datumaro/plugins/yolo_format/extractor.py]: I mean arbitrary strings

Comment (zhiltsov-max) [reply datumaro/plugins/yolo_format/extractor.py]: Could you add an example of the input and supposed parsing result?

Comment (Eldies) [reply datumaro/plugins/yolo_format/extractor.py]: data.yaml has something like that:
```yaml
names:
  0: label0 # <- this is what YOLOv8 supports 
  1: label1  # <- this is what YOLOv8 supports 
  42: another_label  # <- this is allowed by this PR
  string_key: yet_another_label   # <- I suggest to also allow this
```
and then annotation txt file could have something like
```
0 <coordinates>
42 <coordinates>
string_key <coordinates>
```

Keys are dropped while parsing, therefore it is supposed to be parsed to the same dataset as if data.yaml had
```yaml
names:
  0: label0 
  1: label1
  2: another_label 
  3: yet_another_label
```
with annotation txt file
```
0 <coordinates>
2 <coordinates>
3 <coordinates>
```

Comment (zhiltsov-max) [reply datumaro/plugins/yolo_format/extractor.py]: It looks very confusing if a label has 2 names. I don't think arbitrary string keys should be supported, unless we have a clear use case for this.

File: datumaro/plugins/yolo_format/extractor.py [hunk 3]
```diff
@@ -250,6 +250,12 @@ def _parse_annotations(
 
         return annotations
 
+    def _map_label_id(self, label_id):
```
Comment (zhiltsov-max) [inline datumaro/plugins/yolo_format/extractor.py]: Please add type annotations for the new or updated functions.

File: datumaro/plugins/yolo_format/extractor.py [hunk 4]
```diff
@@ -438,6 +463,10 @@ def _check_is_rectangle(p1, p2, p3, p4):
             raise InvalidAnnotationError(
                 "Given points do not form a rectangle: opposite sides have different slope angles."
             )
+        if abs((p12_angle - p23_angle) % math.pi - math.pi / 2) > 0.02:
```
Comment (zhiltsov-max) [inline datumaro/plugins/yolo_format/extractor.py]: Please extract 0.02 into a named constant.

File: datumaro/plugins/yolo_format/extractor.py (hunk 3 above)
Comment (Eldies) [reply datumaro/plugins/yolo_format/extractor.py]: Added

File: datumaro/plugins/yolo_format/extractor.py (hunk 4 above)
Comment (Eldies) [reply datumaro/plugins/yolo_format/extractor.py]: done

File: datumaro/plugins/yolo_format/extractor.py [hunk 5]
```diff
@@ -572,6 +611,13 @@ def make_children_names(skeleton_label):
 
         return categories
 
+    def _map_label_id(self, ann_label_id):
+        skeleton_id = super()._map_label_id(ann_label_id)
+        label_id = self._skeleton_id_to_label_id.get(skeleton_id, -1)
+        if self._categories[AnnotationType.label][label_id].parent != "":
+            raise InvalidAnnotationError("WTF")
```
Comment (zhiltsov-max) [inline datumaro/plugins/yolo_format/extractor.py]: Please make the message more descriptive.

Comment (Eldies) [reply datumaro/plugins/yolo_format/extractor.py]: Removed it. This error is more of a debug artefact, I think this situation can not happen

Comment (zhiltsov-max) [reply datumaro/plugins/yolo_format/extractor.py]: Ok. Typically, an assert is the correct choice in such cases.

**Pull request merged**